### PR TITLE
Add at::cpu namespace of functions for structured kernels

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50958 Skip resize_output call when TensorIterator has proven it's unnecessary
* **#49505 Add at::cpu namespace of functions for structured kernels**

I have a problem which is that static runtime needs a way to bypass
dispatch and call into kernels directly.  Previously, it used
native:: bindings to do this; but these bindings no longer exist
for structured kernels!  Enter at::cpu: a namespace of exactly
at:: compatible functions that assume all of their arguments are
CPU and non-autograd!  The header looks like this:

```
namespace at {
namespace cpu {

CAFFE2_API Tensor & add_out(Tensor & out, const Tensor & self, const Tensor & other, Scalar alpha=1);
CAFFE2_API Tensor add(const Tensor & self, const Tensor & other, Scalar alpha=1);
CAFFE2_API Tensor & add_(Tensor & self, const Tensor & other, Scalar alpha=1);
CAFFE2_API Tensor & upsample_nearest1d_out(Tensor & out, const Tensor & self, IntArrayRef output_size, c10::optional<double> scales=c10::nullopt);
CAFFE2_API Tensor upsample_nearest1d(const Tensor & self, IntArrayRef output_size, c10::optional<double> scales=c10::nullopt);
CAFFE2_API Tensor & upsample_nearest1d_backward_out(Tensor & grad_input, const Tensor & grad_output, IntArrayRef output_size, IntArrayRef input_size, c10::optional<double> scales=c10::nullopt);
CAFFE2_API Tensor upsample_nearest1d_backward(const Tensor & grad_output, IntArrayRef output_size, IntArrayRef input_size, c10::optional<double> scales=c10::nullopt);

}}
```

This slows down static runtime because these are not the "allow
resize of nonzero tensor" variant binding (unlike the ones I had manually
written).  We can restore this: it's a matter of adding codegen smarts to
do this, but I haven't done it just yet since it's marginally more
complicated.

In principle, non-structured kernels could get this treatment too.
But, like an evil mastermind, I'm withholding it from this patch, as an extra
carrot to get people to migrate to structured muahahahaha.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25616105](https://our.internmc.facebook.com/intern/diff/D25616105)

### CODE REVIEW

**Comment by ezyang** [comment]
cc @bwasti @swolchok @ngimel

File: tools/codegen/gen.py [hunk 1]
```diff
@@ -454,6 +461,12 @@ def gen_one(f: NativeFunction) -> Optional[str]:
 }}
 
 }} // anonymous namespace
+
+namespace {self.dispatch_key.lower()} {{
```
Comment (bdhirsh) [inline tools/codegen/gen.py:465]: Probably not a big deal, but the individual `namespace cpu {...}` for every op will probably make these files a few thousand lines longer, vs. grouping them all together in one namespace block

**Review by smessmer — APPROVED** [review]

File: tools/codegen/gen.py (hunk 1 above)
Comment (ezyang) [reply tools/codegen/gen.py:465]: Yeah. The alternative is to split up the implementations in codegen. I'm ambivalent about this; so if someone feels strongly I'll swap it around.

> Actually, isn't this generating at::<dispatch_key> functions for every dispatch key, and then only providing headers for the specific keys we want (cpu/cuda)?

Technically yes, but in reality only CPU and CUDA are supported by structured, so there isn't actually any wastage.

Comment (bdhirsh) [reply tools/codegen/gen.py:465]: > Yeah. The alternative is to split up the implementations in codegen. I'm ambivalent about this; so if someone feels strongly I'll swap it around.

Thought it was worth calling out, but I'm ambivalent as well :)

> Technically yes, but in reality only CPU and CUDA are supported by structured, so there isn't actually any wastage.

Ah right, yeah

File: tools/codegen/gen.py [hunk 2]
```diff
@@ -402,9 +398,20 @@ def gen_one(f: NativeFunction) -> Optional[str]:
...
 
             k = f.func.kind()
-            sig = NativeSignature.from_schema(f.func)
+            sig = NativeSignature(f.func, prefix="wrapper_")
+
+            # Unconditionally generate function version, never fallback binding.
+            # Some extra massaging would then be necessary in a hypothetical
+            # CPUTensor class
+            cpp_sig_group = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False)
+            # For now, don't generate faithful signature for simplicity
```
Comment (bhosmer) [inline tools/codegen/gen.py:407]: Are there any subtle issues involved in also generating faithful sig versions that might be called out here? I can imagine somebody hitting a future use case that wants them (say, binding straight from python to backend-specific functions) and getting caught on something nonobvious, might be worth calling out any such

File: tools/codegen/gen.py [hunk 3]
```diff
@@ -1303,15 +1315,19 @@ def make_file_manager(install_dir: str) -> FileManager:
         # kernels
         "Meta",
     ]
+    # Only a limited set of dispatch keys get CPUFunctions.h headers generated
```
Comment (bhosmer) [inline tools/codegen/gen.py:1318]: s/CPUFunctions.h/{dispatch key}Functions.h/ or whatever

**Review by bhosmer — APPROVED** [review]

File: tools/codegen/gen.py (hunk 2 above)
Comment (ezyang) [reply tools/codegen/gen.py:407]: No, I think we probably should generate faithful version too. I just didn't need it, so I didn't put in the logic for it.

**Comment by ezyang** [comment]
I got papal dispensation from @bwasti to make static runtime a little slower again (this PR brings back `resize({0})` call)

**Pull request closed without merging**